File: probe-go/sqlite.go
// SQLite 수확기 — database/sql의 표준 인터페이스만 사용해 카탈로그 원문을
// 옮긴다. 파싱과 그래프 의미론은 엔진의 책임이므로 이 파일은 PRAGMA와
// sqlite_master가 돌려준 값만 document로 조립한다.

package main

import (
	"database/sql"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

type sqliteObject struct {
	name string
	kind string
	body *string
}

// extractSQLite — SQLite의 attached database마다 하나의 document schema를
// 만든다. temp는 연결 세션에만 존재하므로 네이티브 reader와 같이 건너뛴다.
func (h *harvester) extractSQLite() CatalogDocument {
	schemas := h.sqliteSchemas()
	doc := CatalogDocument{
		Version: documentVersion, Dialect: "sqlite", Reader: "probe-go",
		Schemas: []SchemaDoc{}, Limitations: []string{},
	}
	for _, schema := range schemas {
		doc.Schemas = append(doc.Schemas, h.sqliteSchema(schema))
	}
	sort.Slice(doc.Schemas, func(i, j int) bool { return doc.Schemas[i].Name < doc.Schemas[j].Name })
	doc.Limitations = append(doc.Limitations, h.limitations...)
	return doc
}

func (h *harvester) sqliteSchema(schema string) SchemaDoc {
	objects := h.sqliteObjects(schema)
	sd := SchemaDoc{Name: schema, Objects: []ObjectDoc{}, Routines: []RoutineDoc{}}
	for _, raw := range objects {
		columns := h.sqliteColumns(schema, raw.name)
		constraints := h.sqliteConstraints(schema, raw.name)
		indexes := h.sqliteIndexes(schema, raw.name)
		obj := ObjectDoc{
			Name: raw.name, Kind: raw.kind, Columns: columns,
			Constraints: constraints, Indexes: indexes, Triggers: []TriggerDoc{},
			Body: raw.body,
		}
		for i := range obj.Columns {
			obj.Columns[i].PkPosition = sqlitePrimaryPosition(obj.Columns[i])
		}
		sd.Objects = append(sd.Objects, obj)
	}
	h.sqliteTriggers(schema, sd.Objects)
	sort.Slice(sd.Objects, func(i, j int) bool { return sd.Objects[i].Name < sd.Objects[j].Name })
	return sd
}

func (h *harvester) streamSQLite(stream *ndjsonStreamWriter) error {
	for _, schema := range h.sqliteSchemas() {
		if err := stream.schema(h.sqliteSchema(schema)); err != nil {
			return err
		}
		if err := h.streamDependencies(stream, schema); err != nil {
			return err
		}
	}
	return nil
}

// sqliteSchemas — database_list의 attached 순서를 정렬해 결과를 결정적으로 만든다.
func (h *harvester) sqliteSchemas() []string {
	set := map[string]bool{}
	h.bestEffort("sqlite schemas", "PRAGMA database_list", func(rows *sql.Rows) error {
		v, err := sqliteRow(rows)
		if err != nil {
			return err
		}
		name := sqliteString(v, "name")
		if name != "temp" && (len(h.schemaFilter) == 0 || h.schemaFilter[name]) {
			set[name] = true
		}
		return nil
	})
	out := make([]string, 0, len(set))
	for name := range set {
		out = append(out, name)
	}
	sort.Strings(out)
	return out
}

// sqliteObjects — sqlite_master의 table/view만 객체 정점으로 만든다. 인덱스와
// trigger는 각각 PRAGMA와 객체 연결 단계에서 수확한다.
func (h *harvester) sqliteObjects(schema string) []sqliteObject {
	objects := []sqliteObject{}
	inlineConstraints := false
	query := fmt.Sprintf(`SELECT name, type, sql, tbl_name FROM %s.sqlite_master
                         WHERE name NOT LIKE 'sqlite_%%' ORDER BY name`, sqliteQuoteIdent(schema))
	h.bestEffort("sqlite objects", query, func(rows *sql.Rows) error {
		v, err := sqliteRow(rows)
		if err != nil {
			return err
		}
		name, kind := sqliteString(v, "name"), sqliteString(v, "type")
		body := sqliteStringPtr(v, "sql")
		if body != nil {
			lower := strings.ToLower(*body)
			inlineConstraints = inlineConstraints || strings.Contains(lower, "unique") || strings.Contains(lower, "check")
		}
		switch kind {
		case "table", "view":
			objects = append(objects, sqliteObject{
				name: name, kind: kind, body: body,
			})
		}
		return nil
	})
	if inlineConstraints {
		h.catalogIncomplete = true
		h.limitations = append(h.limitations,
			"inline UNIQUE/CHECK constraints are not exposed by the catalog; inline DDL constraint parsing is unsupported")
	}
	sort.Slice(objects, func(i, j int) bool { return objects[i].name < objects[j].name })
	return objects
}

// sqliteColumns — table_xinfo는 generated/hidden 컬럼도 포함하므로 ordinary
// table_info보다 네이티브 reader의 컬럼 집합과 맞는다.
func (h *harvester) sqliteColumns(schema, table string) []ColumnDoc {
	columns := []ColumnDoc{}
	query := fmt.Sprintf("PRAGMA %s.table_xinfo(%s)", sqliteQuoteIdent(schema), sqliteQuoteIdent(table))
	h.bestEffort("sqlite columns", query, func(rows *sql.Rows) error {
		v, err := sqliteRow(rows)
		if err != nil {
			return err
		}
		pk := sqliteInt(v, "pk")
		columns = append(columns, ColumnDoc{
			Name: sqliteString(v, "name"), DataType: sqliteString(v, "type"),
			Nullable: sqliteInt(v, "notnull") == 0 && pk == 0,
			Default:  sqliteStringPtr(v, "dflt_value"), Ordinal: len(columns) + 1,
			PkPosition: int(pk),
		})
		return nil
	})
	return columns
}

// sqliteConstraints — foreign_key_list의 id로 다중 컬럼 FK를 묶고,
// table_xinfo의 pk 위치로 SQLite가 이름을 주지 않는 PK를 복원한다.
func (h *harvester) sqliteConstraints(schema, table string) []ConstraintDoc {
	type fkPart struct {
		seq  int
		from string
		to   string
	}
	groups := map[int][]fkPart{}
	targets := map[int]string{}
	query := fmt.Sprintf("PRAGMA %s.foreign_key_list(%s)", sqliteQuoteIdent(schema), sqliteQuoteIdent(table))
	h.bestEffort("sqlite foreign keys", query, func(rows *sql.Rows) error {
		v, err := sqliteRow(rows)
		if err != nil {
			return err
		}
		id := int(sqliteInt(v, "id"))
		targets[id] = sqliteString(v, "table")
		groups[id] = append(groups[id], fkPart{
			seq: int(sqliteInt(v, "seq")), from: sqliteString(v, "from"), to: sqliteString(v, "to"),
		})
		return nil
	})
	constraints := []ConstraintDoc{}
	ids := make([]int, 0, len(groups))
	for id := range groups {
		ids = append(ids, id)
	}
	sort.Ints(ids)
	for _, id := range ids {
		parts := groups[id]
		sort.Slice(parts, func(i, j int) bool { return parts[i].seq < parts[j].seq })
		columns, referenced := []string{}, []string{}
		for _, part := range parts {
			columns = append(columns, part.from)
			referenced = append(referenced, part.to)
		}
		constraints = append(constraints, ConstraintDoc{
			Name: fmt.Sprintf("%s_fk_%d", table, id), Kind: "fk", Columns: columns,
			Referenced: &ReferencedDoc{Table: targets[id], Columns: referenced},
		})
	}

	pkColumns := []ColumnDoc{}
	for _, column := range h.sqliteColumns(schema, table) {
		if column.PkPosition > 0 {
			pkColumns = append(pkColumns, column)
		}
	}
	sort.Slice(pkColumns, func(i, j int) bool { return pkColumns[i].PkPosition < pkColumns[j].PkPosition })
	if len(pkColumns) > 0 {
		columns := make([]string, 0, len(pkColumns))
		for _, column := range pkColumns {
			columns = append(columns, column.Name)
		}
		constraints = append(constraints, ConstraintDoc{
			Name: table + "_pk", Kind: "pk", Columns: columns,
		})
	}
	sort.Slice(constraints, func(i, j int) bool { return constraints[i].Name < constraints[j].Name })
	return constraints
}

// sqliteIndexes — index_xinfo의 rowid/표현식 항목(cid < 0)은 컬럼 이름이
// 없으므로 정점은 보존하되 컬럼 목록에서는 제외한다.
func (h *harvester) sqliteIndexes(schema, table string) []IndexDoc {
	type indexRow struct {
		name    string
		unique  bool
		partial bool
	}
	indexes := []indexRow{}
	listQuery := fmt.Sprintf("PRAGMA %s.index_list(%s)", sqliteQuoteIdent(schema), sqliteQuoteIdent(table))
	h.bestEffort("sqlite indexes", listQuery, func(rows *sql.Rows) error {
		v, err := sqliteRow(rows)
		if err != nil {
			return err
		}
		name := sqliteString(v, "name")
		if name != "" {
			indexes = append(indexes, indexRow{name: name, unique: sqliteInt(v, "unique") == 1, partial: sqliteInt(v, "partial") == 1})
		}
		return nil
	})
	out := make([]IndexDoc, 0, len(indexes))
	for _, index := range indexes {
		columns := []string{}
		complete := true
		notesBefore := len(h.limitations)
		query := fmt.Sprintf("PRAGMA %s.index_xinfo(%s)", sqliteQuoteIdent(schema), sqliteQuoteIdent(index.name))
		h.bestEffort("sqlite index columns", query, func(rows *sql.Rows) error {
			v, err := sqliteRow(rows)
			if err != nil {
				return err
			}
			if sqliteInt(v, "key") == 0 {
				return nil
			}
			if sqliteInt(v, "cid") >= 0 && sqliteString(v, "name") != "" {
				if name := sqliteString(v, "name"); name != "" {
					columns = append(columns, name)
				}
			} else {
				complete = false
			}
			return nil
		})
		out = append(out, IndexDoc{Name: index.name, Unique: index.unique, Columns: columns, DefinitionComplete: boolValue(complete && len(columns) > 0 && len(h.limitations) == notesBefore), HasPredicate: boolValue(index.partial)})
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Name < out[j].Name })
	return out
}

// sqliteTriggers — master 행은 trigger의 tbl_name만 주므로 이미 수확한
// 객체에 이름으로 귀속한다. 대상이 없으면 유령 정점을 만들지 않고 limitation만 남긴다.
func (h *harvester) sqliteTriggers(schema string, objects []ObjectDoc) {
	byName := map[string]*ObjectDoc{}
	for i := range objects {
		byName[objects[i].Name] = &objects[i]
	}
	query := fmt.Sprintf(`SELECT name, tbl_name, sql FROM %s.sqlite_master
                         WHERE type = 'trigger' AND name NOT LIKE 'sqlite_%%' ORDER BY name`, sqliteQuoteIdent(schema))
	h.bestEffort("sqlite triggers", query, func(rows *sql.Rows) error {
		v, err := sqliteRow(rows)
		if err != nil {
			return err
		}
		table := sqliteString(v, "tbl_name")
		obj := byName[table]
		if obj == nil {
			h.catalogIncomplete = true
			h.limitations = append(h.limitations, fmt.Sprintf("trigger %s.%s targets %s, which is missing from the scanned catalog", schema, sqliteString(v, "name"), table))
			return nil
		}
		obj.Triggers = append(obj.Triggers, TriggerDoc{Name: sqliteString(v, "name"), Body: sqliteStringPtr(v, "sql")})
		return nil
	})
	for _, obj := range objects {
		sort.Slice(obj.Triggers, func(i, j int) bool { return obj.Triggers[i].Name < obj.Triggers[j].Name })
	}
}

// sqliteQuoteIdent는 PRAGMA에도 사용할 수 있는 SQLite 식별자 인용이다.
// 입력을 SQL 조각으로 취급하지 않고 큰따옴표를 두 배로 만들어 주입을 막는다.
func sqliteQuoteIdent(name string) string {
	return `"` + strings.ReplaceAll(name, `"`, `""`) + `"`
}

// sqliteRow는 PRAGMA 버전별 추가 컬럼을 허용하기 위해 열 이름으로 한 행을
// 읽는다. database/sql.Rows는 반드시 이 함수가 끝난 뒤 닫히므로 단일 연결
// 풀에서도 다음 메타데이터 질의가 기다리지 않는다.
func sqliteRow(rows *sql.Rows) (map[string]any, error) {
	columns, err := rows.Columns()
	if err != nil {
		return nil, err
	}
	values := make([]any, len(columns))
	dest := make([]any, len(columns))
	for i := range values {
		dest[i] = &values[i]
	}
	if err := rows.Scan(dest...); err != nil {
		return nil, err
	}
	row := make(map[string]any, len(columns))
	for i, name := range columns {
		row[name] = values[i]
	}
	return row, nil
}

func sqliteValue(row map[string]any, name string) any {
	if value, ok := row[name]; ok {
		return value
	}
	for key, value := range row {
		if strings.EqualFold(key, name) {
			return value
		}
	}
	return nil
}

func sqliteString(row map[string]any, name string) string {
	switch value := sqliteValue(row, name).(type) {
	case string:
		return value
	case []byte:
		return string(value)
	default:
		return ""
	}
}

func sqliteStringPtr(row map[string]any, name string) *string {
	if sqliteValue(row, name) == nil {
		return nil
	}
	value := sqliteString(row, name)
	return &value
}

func sqliteInt(row map[string]any, name string) int64 {
	switch value := sqliteValue(row, name).(type) {
	case int64:
		return value
	case int:
		return int64(value)
	case int32:
		return int64(value)
	case uint64:
		return int64(value)
	case []byte:
		parsed, _ := strconv.ParseInt(string(value), 10, 64)
		return parsed
	case string:
		parsed, _ := strconv.ParseInt(value, 10, 64)
		return parsed
	default:
		return 0
	}
}

func sqlitePrimaryPosition(column ColumnDoc) int {
	return column.PkPosition
}

File: engine/source/src/sqlite.rs
//! SQLite 네이티브 reader — sqlx로 카탈로그를 읽어 CatalogDocument를 만든다.
//!
//! PRAGMA 계열은 파라미터 바인딩이 안 되므로 식별자를 직접 이스케이프해
//! 포맷한다. 읽기 전용(`mode=ro`)으로 연다 — 스캐너가 DB를 바꾸는 일은
//! 없어야 한다.

use sqlx::sqlite::{SqliteConnectOptions, SqlitePool, SqlitePoolOptions};
use sqlx::Row;
use std::str::FromStr;

use crate::document::*;
use crate::SourceError;

/// `sqlite:경로`·`sqlite://경로`·베어 파일 경로를 받아 카탈로그를 읽는다.
pub async fn read(url: &str) -> Result<CatalogDocument, SourceError> {
    let path = strip_sqlite_url(url);
    let options = SqliteConnectOptions::from_str(&format!("sqlite:{path}"))
        .map_err(|e| SourceError::Connect(format!("sqlite 경로 해석 실패: {e}")))?
        .read_only(true);
    let pool = SqlitePoolOptions::new()
        .connect_with(options)
        .await
        .map_err(|e| SourceError::Connect(format!("sqlite 접속 실패 ({path}): {e}")))?;

    let mut doc = CatalogDocument {
        context: None,
        dependencies: Vec::new(),
        version: DOCUMENT_VERSION,
        dialect: "sqlite".to_owned(),
        reader: "native-sqlx".to_owned(),
        schemas: Vec::new(),
        limitations: Vec::new(),
    };

    let schema_names = database_list(&pool).await?;
    for schema in schema_names {
        doc.schemas
            .push(read_schema(&pool, &schema, &mut doc.limitations).await?);
    }
    doc.schemas.sort_by(|a, b| a.name.cmp(&b.name));
    doc.context = Some(CollectionContext {
        source_id: String::new(),
        database: Some("main".into()),
        schema_filter: None,
        catalog_complete: doc.limitations.is_empty(),
    });
    Ok(doc)
}

/// `sqlite:`/`sqlite://` 접두와 쿼리스트링을 벗긴다.
fn strip_sqlite_url(url: &str) -> String {
    let body = url
        .strip_prefix("sqlite://")
        .or_else(|| url.strip_prefix("sqlite:"))
        .unwrap_or(url);
    body.split('?').next().unwrap_or(body).to_owned()
}

/// attached database 목록. temp는 스캔하지 않는다 — 임시 객체는 세션 한정이라
/// 의존성 분석의 대상이 아니다.
async fn database_list(pool: &SqlitePool) -> Result<Vec<String>, SourceError> {
    let rows = sqlx::query("PRAGMA database_list")
        .fetch_all(pool)
        .await
        .map_err(SourceError::Query)?;
    Ok(rows
        .iter()
        .map(|r| r.get::<String, _>("name"))
        .filter(|n| n != "temp")
        .collect())
}

/// PRAGMA에 넣을 식별자 이스케이프. 큰따옴표를 두 번 쓴다.
fn qi(name: &str) -> String {
    format!("\"{}\"", name.replace('"', "\"\""))
}

async fn read_schema(
    pool: &SqlitePool,
    schema: &str,
    limitations: &mut Vec<String>,
) -> Result<SchemaDoc, SourceError> {
    // sqlite_master는 테이블·뷰·트리거·인덱스를 한 데 담는다.
    // sqlite_% 접두는 내부 객체(autoincrement 시퀀스 등)라 제외한다.
    let rows = sqlx::query(&format!(
        "SELECT name, type, sql, tbl_name FROM {schema}.sqlite_master \
         WHERE name NOT LIKE 'sqlite_%' ORDER BY name",
        schema = qi(schema)
    ))
    .fetch_all(pool)
    .await
    .map_err(SourceError::Query)?;

    let mut objects: Vec<ObjectDoc> = Vec::new();
    let mut pending_triggers: Vec<(String, String, Option<String>)> = Vec::new();
    let mut inline_constraints = false;

    for row in &rows {
        let (name, kind, sql, tbl_name): (String, String, Option<String>, Option<String>) = (
            row.get("name"),
            row.get("type"),
            row.get("sql"),
            row.get("tbl_name"),
        );
        match kind.as_str() {
            "table" | "view" => {
                // UNIQUE/CHECK는 DDL 안에 박혀 있어 PRAGMA로 못 읽는다.
                // 엔진의 몸체 파서가 보강할 때까지 한계로 남긴다.
                if let Some(ddl) = &sql {
                    let lower = ddl.to_lowercase();
                    if lower.contains("unique") || lower.contains("check") {
                        inline_constraints = true;
                    }
                }
                objects.push(ObjectDoc {
                    name: name.clone(),
                    kind: if kind == "table" { "table" } else { "view" }.to_owned(),
                    columns: read_columns(pool, schema, &name).await?,
                    constraints: read_constraints(pool, schema, &name).await?,
                    indexes: read_indexes(pool, schema, &name).await?,
                    triggers: Vec::new(),
                    body: sql,
                    // SQLite는 사용 통계 카탈로그가 없다 — 미수집과 0 관측을 구분해 None.
                    usage: None,
                });
            }
            "trigger" => {
                if let Some(tbl) = tbl_name {
                    pending_triggers.push((tbl, name, sql));
                }
            }
            // autoindex·수동 인덱스는 index_list에서 컬럼 정보와 함께 읽는다.
            _ => {}
        }
    }

    for (tbl, name, sql) in pending_triggers {
        if let Some(obj) = objects.iter_mut().find(|o| o.name == tbl) {
            obj.triggers.push(TriggerDoc { name, body: sql });
        } else {
            limitations.push(format!(
                "trigger {schema}.{name}의 대상 {tbl}을 스캔에서 찾지 못함"
            ));
        }
    }

    if inline_constraints {
        limitations.push(
            "DDL 인라인 UNIQUE/CHECK 제약은 카탈로그에 없어 수집하지 못함 (인라인 제약 파싱 미지원)"
                .to_owned(),
        );
    }

    objects.sort_by(|a, b| a.name.cmp(&b.name));
    for obj in &mut objects {
        obj.columns.sort_by_key(|c| c.ordinal);
        obj.constraints.sort_by(|a, b| a.name.cmp(&b.name));
        obj.indexes.sort_by(|a, b| a.name.cmp(&b.name));
        obj.triggers.sort_by(|a, b| a.name.cmp(&b.name));
    }
    Ok(SchemaDoc {
        name: schema.to_owned(),
        objects,
        routines: Vec::new(),
    })
}

async fn read_columns(
    pool: &SqlitePool,
    schema: &str,
    table: &str,
) -> Result<Vec<ColumnDoc>, SourceError> {
    // table_xinfo는 hidden/generated 컬럼까지 보여준다.
    let rows = sqlx::query(&format!(
        "PRAGMA {schema}.table_xinfo({table})",
        schema = qi(schema),
        table = qi(table)
    ))
    .fetch_all(pool)
    .await
    .map_err(SourceError::Query)?;
    Ok(rows
        .iter()
        .enumerate()
        .map(|(i, r)| ColumnDoc {
            name: r.get::<String, _>("name"),
            data_type: r.get::<String, _>("type"),
            nullable: r.get::<i64, _>("notnull") == 0 && r.get::<i64, _>("pk") == 0,
            default: r.get::<Option<String>, _>("dflt_value"),
            ordinal: (i + 1) as u32,
            pk_position: r.get::<i64, _>("pk") as u32,
        })
        .collect())
}

async fn read_constraints(
    pool: &SqlitePool,
    schema: &str,
    table: &str,
) -> Result<Vec<ConstraintDoc>, SourceError> {
    let mut constraints = Vec::new();

    // FK 목록. 여러 컬럼 FK는 같은 id로 묶인다.
    let fk_rows = sqlx::query(&format!(
        "PRAGMA {schema}.foreign_key_list({table})",
        schema = qi(schema),
        table = qi(table)
    ))
    .fetch_all(pool)
    .await
    .map_err(SourceError::Query)?;

    let mut fk_groups: std::collections::BTreeMap<i64, Vec<(String, String)>> =
        std::collections::BTreeMap::new();
    let mut fk_targets: std::collections::BTreeMap<i64, String> = std::collections::BTreeMap::new();
    for r in &fk_rows {
        let id = r.get::<i64, _>("id");
        fk_targets.insert(id, r.get::<String, _>("table"));
        fk_groups.entry(id).or_default().push((
            r.get::<String, _>("from"),
            r.get::<Option<String>, _>("to").unwrap_or_default(),
        ));
    }
    for (id, pairs) in fk_groups {
        constraints.push(ConstraintDoc {
            name: format!("{table}_fk_{id}"),
            kind: "fk".to_owned(),
            columns: pairs.iter().map(|(from, _)| from.clone()).collect(),
            referenced: Some(ReferencedDoc {
                // SQLite의 foreign_key_list는 대상 스키마를 안 알려준다 —
                // 같은 스키마로 단정하지 않고 None으로 둔다.
                schema: None,
                table: fk_targets[&id].clone(),
                columns: pairs.iter().map(|(_, to)| to.clone()).collect(),
            }),
        });
    }

    // PK는 table_xinfo의 pk 위치로 재구성한다(SQLite는 제약 이름이 없다).
    let pk_cols = read_columns(pool, schema, table)
        .await?
        .into_iter()
        .filter(|c| c.pk_position > 0)
        .map(|c| (c.pk_position, c.name))
        .collect::<Vec<_>>();
    if !pk_cols.is_empty() {
        let mut pk_cols = pk_cols;
        pk_cols.sort();
        constraints.push(ConstraintDoc {
            name: format!("{table}_pk"),
            kind: "pk".to_owned(),
            columns: pk_cols.into_iter().map(|(_, n)| n).collect(),
            referenced: None,
        });
    }
    Ok(constraints)
}

async fn read_indexes(
    pool: &SqlitePool,
    schema: &str,
    table: &str,
) -> Result<Vec<IndexDoc>, SourceError> {
    let list = sqlx::query(&format!(
        "PRAGMA {schema}.index_list({table})",
        schema = qi(schema),
        table = qi(table)
    ))
    .fetch_all(pool)
    .await
    .map_err(SourceError::Query)?;

    let mut indexes = Vec::new();
    for r in &list {
        let idx_name: String = r.get("name");
        let cols = sqlx::query(&format!(
            "PRAGMA {schema}.index_xinfo({idx})",
            schema = qi(schema),
            idx = qi(&idx_name)
        ))
        .fetch_all(pool)
        .await
        .map_err(SourceError::Query)?;
        // key=0은 WITHOUT ROWID 보조 키라 인덱스 prefix가 아니다. key=1인
        // 식/표현식 항목은 이름을 복원할 수 없으므로 definition을 불완전하게 둔다.
        let mut definition_complete = true;
        let columns: Vec<String> = cols
            .iter()
            .filter_map(|c| {
                if c.get::<i64, _>("key") == 0 {
                    return None;
                }
                let cid: i64 = c.get("cid");
                if cid < 0 {
                    definition_complete = false;
                    return None;
                }
                match c.get::<Option<String>, _>("name") {
                    Some(name) => Some(name),
                    None => {
                        definition_complete = false;
                        None
                    }
                }
            })
            .collect();
        let partial = r.get::<i64, _>("partial") == 1;
        indexes.push(IndexDoc {
            has_predicate: Some(partial),
            definition_complete: Some(definition_complete),
            predicate: None,
            name: idx_name,
            unique: r.get::<i64, _>("unique") == 1,
            columns,
            usage: None,
        });
    }
    Ok(indexes)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// fixture SQL을 적용한 임시 DB를 만들어 document를 읽는다.
    async fn fixture_doc(sql: &str) -> CatalogDocument {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("t.db");
        let pool = SqlitePoolOptions::new()
            .connect(&format!("sqlite:{}?mode=rwc", path.display()))
            .await
            .unwrap();
        // 트리거 몸체의 BEGIN..END 안에도 ';'가 있으므로 naive split은 못 쓴다 —
        // raw_sql이 문 단위로 나눠 실행해 준다.
        sqlx::raw_sql(sql).execute(&pool).await.unwrap();
        pool.close().await;
        let doc = read(&format!("sqlite:{}", path.display())).await.unwrap();
        // tempdir이 drop돼도 doc은 이미 메모리에 있다.
        doc
    }

    #[tokio::test]
    async fn 테이블과_fk와_pk를_읽는다() {
        let doc = fixture_doc(
            "CREATE TABLE customers(id INTEGER PRIMARY KEY, name TEXT);
             CREATE TABLE orders(id INTEGER PRIMARY KEY, customer_id INTEGER NOT NULL REFERENCES customers(id))",
        )
        .await;
        let main = &doc.schemas[0];
        assert_eq!(main.name, "main");
        let orders = main.objects.iter().find(|o| o.name == "orders").unwrap();
        let fk = orders.constraints.iter().find(|c| c.kind == "fk").unwrap();
        assert_eq!(fk.columns, ["customer_id"]);
        let referenced = fk.referenced.as_ref().unwrap();
        assert_eq!(referenced.table, "customers");
        assert_eq!(referenced.columns, ["id"]);
        assert!(orders.constraints.iter().any(|c| c.kind == "pk"));
    }

    #[tokio::test]
    async fn view_본문을_수집한다() {
        let doc = fixture_doc(
            "CREATE TABLE a(id INTEGER PRIMARY KEY);
             CREATE VIEW v AS SELECT id FROM a",
        )
        .await;
        let view = doc.schemas[0]
            .objects
            .iter()
            .find(|o| o.name == "v")
            .unwrap();
        assert_eq!(view.kind, "view");
        assert!(view.body.as_ref().unwrap().contains("SELECT"));
    }

    #[tokio::test]
    async fn 트리거는_대상_테이블에_붙는다() {
        let doc = fixture_doc(
            "CREATE TABLE a(id INTEGER PRIMARY KEY);
             CREATE TRIGGER t1 AFTER INSERT ON a BEGIN SELECT 1; END",
        )
        .await;
        let a = doc.schemas[0]
            .objects
            .iter()
            .find(|o| o.name == "a")
            .unwrap();
        assert_eq!(a.triggers.len(), 1);
        assert_eq!(a.triggers[0].name, "t1");
    }
}
